subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskOutputs.java

```diff
@@ -200,6 +200,12 @@ public FileCollection getFiles() {
                     if (propertySpec instanceof CompositeTaskOutputPropertySpec) {
                         return ((CompositeTaskOutputPropertySpec) propertySpec).resolveToOutputProperties();
                     } else {
+                        if (propertySpec instanceof CacheableTaskOutputFilePropertySpec) {
+                            File outputFile = ((CacheableTaskOutputFilePropertySpec) propertySpec).getOutputFile();
+                            if (outputFile == null) {
+                                return Iterators.emptyIterator();
+                            }
+                        }
```

Comment:
What happens when we have `@OutputFiles List<File>` that goes from `[null, File]` to `[File, File]`?

What happens when it goes from `[null, File]` to `[File, null]`?

Reply:
We cannot have `null` in the file collection - we can have a provide returning `null`. If this is the case, then `null` will be ignored. So your cases break down to:
```
before                         | after
['outputFile1']                | ['outputFile1', 'outputFile2']
['outputFile1', 'outputFile2'] | ['outputFile1']
[]                             | ['outputFile1', 'outputFile2']
['outputFile1']                | ['outputFile2']
```
All these changes go undetected :(
In order to detect these changes, we would need compare the declared output paths from the last execution to the declared output paths of the current execution.
Do you have another idea? Do we want to fix this problem, too?

Reply:
I think it's fine to leave the plural outputs as they are, but we should make the behavior explicit. What `@OutputFiles` on an `Iterable` means is that the task will produce a number of output files, but we do not know the precise identity of these files (whereas we do for `Map`s). In such a case we can as well ignore the ordering of those outputs.

This means that:

`[null, File]` is actually the same as `[File, null]` which in turn is the same as `[File]`.

I think all we need here is an integration test that demonstrates this behavior.

Reply:
BTW, why can't we have `null` in the collection?

Reply:
We cannot have null in the collection since we resolve each object in the file collection by the file resolver. The file resolver throws an exception when trying to resolve `null`.

Reply:
I added a test.